rotors_gazebo_plugins/src/gazebo_mavlink_interface.cpp

```diff
@@ -0,0 +1,236 @@
+ * Copyright 2015 Michael Burri, ASL, ETH Zurich, Switzerland
+ * Copyright 2015 Mina Kamel, ASL, ETH Zurich, Switzerland
+ * Copyright 2015 Janosch Nikolic, ASL, ETH Zurich, Switzerland
+ * Copyright 2015 Markus Achtelik, ASL, ETH Zurich, Switzerland
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+#include "rotors_gazebo_plugins/gazebo_mavlink_interface.h"
+
+namespace gazebo {
+
+GazeboMavlinkInterface::~GazeboMavlinkInterface() {
+  event::Events::DisconnectWorldUpdateBegin(updateConnection_);
+  if (node_handle_) {
+    node_handle_->shutdown();
+    delete node_handle_;
+  }
+}
+
+void GazeboMavlinkInterface::Load(physics::ModelPtr _model, sdf::ElementPtr _sdf) {
+  // Store the pointer to the model.
+  model_ = _model;
+
+  world_ = model_->GetWorld();
+
+  namespace_.clear();
+
+  if (_sdf->HasElement("robotNamespace"))
+    namespace_ = _sdf->GetElement("robotNamespace")->Get<std::string>();
+  else
+    gzerr << "[gazebo_mavlink_interface] Please specify a robotNamespace.\n";
+
+  node_handle_ = new ros::NodeHandle(namespace_);
+
+  getSdfParam<std::string>(_sdf, "motorSpeedCommandPubTopic", motor_velocity_reference_pub_topic_,
+                           motor_velocity_reference_pub_topic_);
+
+  // Listen to the update event. This event is broadcast every
+  // simulation iteration.
+  updateConnection_ = event::Events::ConnectWorldUpdateBegin(
+      boost::bind(&GazeboMavlinkInterface::OnUpdate, this, _1));
+
+  mav_control_sub_ = node_handle_->subscribe(mavlink_control_sub_topic_, 10,
+                                           &GazeboMavlinkInterface::MavlinkControlCallback,
+                                           this);
+  // Subscriber to IMU sensor_msgs::Imu Message.
+  imu_sub_ = node_handle_->subscribe(imu_sub_topic_, 10, &GazeboMavlinkInterface::ImuCallback, this);
+  
+  motor_velocity_reference_pub_ = node_handle_->advertise<mav_msgs::CommandMotorSpeed>(motor_velocity_reference_pub_topic_, 10);
+  hil_sensor_pub_ = node_handle_->advertise<mavros_msgs::Mavlink>(hil_sensor_mavlink_pub_topic_, 10);
+
+  _rotor_count = 4;
+  last_time_ = world_->GetSimTime();
+  last_gps_time_ = world_->GetSimTime();
+  double gps_update_interval_ = 200*1000000;  // nanoseconds for 5Hz
+
+  gravity_W_ = world_->GetPhysicsEngine()->GetGravity();
+
+  // Magnetic field data for Zurich from WMM2015 (10^5xnanoTesla (N, E, D))
+  mag_W_ = {0.21523, 0.00771, 0.42741};
```

Comment:
Please set all these params from constants in the header so that it's obvious how to change these/what these are.